### PR TITLE
[883/914/879] Add/update API fields for work/volunteering experience

### PR DESCRIPTION
**Pull request by elliotcm**
## Context

This brings the API in line with Manage.

### Added `skills_relevant_to_teaching`

Screenshot 2022-11-21 at 13 54 33
Screenshot 2022-11-21 at 13 53 26

### Added new `start_month`/`end_month` and deprecated `start_date`/`end_date`

Screenshot 2022-11-22 at 13 45 47
Screenshot 2022-11-22 at 13 45 59
Screenshot 2022-11-22 at 13 45 40
Screenshot 2022-11-22 at 13 55 46

### Marked `description` as optional
Screenshot 2022-11-22 at 14 09 07
Screenshot 2022-11-22 at 14 10 01

### Updated the draft spec and documentation for reference changes

Screenshot 2022-11-22 at 15 58 36
Screenshot 2022-11-22 at 15 58 45

## Things to check

- [x] API release notes have been updated if necessary

### CODE REVIEW

**Comment by elliotcm** [comment]
[API 1.3: Add fields relating to work history that are currently missing](https://trello.com/c/JJ7F00XO/883-api-13-add-fields-relating-to-work-history-that-are-currently-missing)

**Comment by elliotcm** [comment]
[Update API reference doc with 1.2 changes](https://trello.com/c/JEuMbmWS/914-update-api-reference-doc-with-12-changes)

**Comment by elliotcm** [comment]
[Write the API documentation for reference field nullification](https://trello.com/c/Bbax3UvE/879-write-the-api-documentation-for-reference-field-nullification)

File: app/presenters/api_docs/api_schema.rb [hunk 1]
```diff
@@ -3,14 +3,26 @@ class APISchema
...
+    def self.null_types(schema)
+      all_types(schema).select { |type| type['type'].nil? }
+    end
+
+    def self.non_null_types(schema)
+      all_types(schema).reject { |type| type['type'].nil? }
+    end
+
+    def self.all_types(schema)
+      (schema['anyOf']&.to_a || []) + (schema['oneOf']&.to_a || []) + (schema['allOf']&.to_a || [])
```
Comment (tomas-stefano) [inline app/presenters/api_docs/api_schema.rb:15]: On this particular case can we do it?

```
  Array(schema['anyOf']) + Array(schema['oneOf']) + Array(schema['allOf'])
```

**Review by tomas-stefano — APPROVED** [review]

File: app/presenters/api_docs/api_schema.rb (hunk 1 above)
Comment (elliotcm) [reply app/presenters/api_docs/api_schema.rb]: Yep that works, done :)

**Comment by frankieroberto** [comment]
Is it worth marking `estimated` as optional? Even though it's currently required (true or false), there are some legacy applications where that question wasn't asked, and it's also possible that we’ll remove that field again in future...

**Comment by elliotcm** [comment]
@frankieroberto 
> Is it worth marking `estimated` as optional? Even though it's currently required (true or false), there are some legacy applications where that question wasn't asked, and it's also possible that we’ll remove that field again in future...

"Optional" in this case means the value can be `null`, so in the case where there's legit `null`s in the database then we should do that, yeah.

If we remove the field in the future that's a different matter, as the "optional" part won't apply, we'd need to deprecate it and remove it as a normal breaking change.

**Comment by frankieroberto** [comment]
> "Optional" in this case means the value can be `null`, so in the case where there's legit `null`s in the database then we should do that, yeah.
> 
> If we remove the field in the future that's a different matter, as the "optional" part won't apply, we'd need to deprecate it and remove it as a normal breaking change.

Fair point. I was thinking about whether we build in a bit more forward-compatibility by encouraging API-users not to build their implementations in such a way that things break if fields start returning nulls (or get removed)...

**Comment by elliotcm** [comment]
> > "Optional" in this case means the value can be `null`, so in the case where there's legit `null`s in the database then we should do that, yeah.
> > If we remove the field in the future that's a different matter, as the "optional" part won't apply, we'd need to deprecate it and remove it as a normal breaking change.
> 
> Fair point. I was thinking about whether we build in a bit more forward-compatibility by encouraging API-users not to build their implementations in such a way that things break if fields start returning nulls (or get removed)...

Understandable, but really that's the nature of the API beast.  Gotta have stability and rules or nobody will use it.

**Review by JR-G — APPROVED** [review]

**Pull request merged**